eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Deleg.hs

```diff
@@ -0,0 +1,255 @@
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UndecidableInstances #-}
+{-# LANGUAGE EmptyDataDeriving #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# OPTIONS_GHC -Wno-orphans #-}
+
+module Cardano.Ledger.Conway.Rules.Deleg (
+  ConwayDELEG,
+  ConwayDelegEvent (..),
+  ConwayDelegPredFailure (..),
+)
+where
+
+import Cardano.Ledger.Conway.TxCert (
+  ConwayTxCert (ConwayTxCertDeleg),
+  ConwayDelegCert (ConwayRegCert, ConwayUnRegCert, ConwayDelegCert, ConwayRegDelegCert), Delegatee (DelegStake, DelegVote, DelegStakeVote),
+ )
+import Cardano.Ledger.Conway.Era (ConwayDELEG)
+import Cardano.Ledger.BaseTypes (
+  ShelleyBase,
+  maybeToStrictMaybe, 
+  StrictMaybe (SNothing),
+ )
+import Cardano.Ledger.Shelley.Rules (DelegEnv(DelegEnv))
+import Cardano.Ledger.Shelley.LedgerState (
+  DState (..),
+ )
+import Control.State.Transition (
+  STS,
+  TRC (TRC),
+  State,
+  Signal,
+  Environment,
+  BaseM,
+  Event,
+  TransitionRule,
+  PredicateFailure,
+  transitionRules,
+  judgmentContext, 
+  (?!), 
+  failBecause,
+ )
+import GHC.Generics (Generic)
+import NoThunks.Class (NoThunks (..))
+import Cardano.Ledger.Shelley.API (CertState(CertState, certDState, certPState, certVState))
+import Cardano.Ledger.Credential (Credential)
+import Cardano.Ledger.Keys (KeyRole(Staking, Voting))
+import Cardano.Ledger.BaseTypes (StrictMaybe(SJust))
+import qualified Cardano.Ledger.UMap as UM
+import Lens.Micro ((^.))
+import Cardano.Ledger.Coin (Coin)
+import Cardano.Ledger.CertState (VState (VState, vsDReps))
+import qualified Data.Set as Set
+import Cardano.Ledger.UMap (UView(RewDepUView, DRepUView, SPoolUView))
+import Cardano.Ledger.Core (Era(EraCrypto), EraRule, EraTx, TxCert)
+import Cardano.Ledger.Conway.Core ({-ShelleyEraTxBody,-} ConwayEraTxBody, ppKeyDepositL)
+import qualified Data.Map.Strict as Map
+
+data ConwayDelegPredFailure era
+  = RegCertProvidedCoinNotEqualToExpectedPoolDepositDELEG
```

Comment:
Maybe `IncorrectDepositProvided` would be a better name.

Reply:
Also it should be fine to use the same error for all of the certificates, having a copy of the same error for each certificate is a pain to maintain.

Reply:
Got it! :raised_hands: